spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/functions/TruncateFunction.java

```diff
@@ -0,0 +1,395 @@
+    }
+
+    @Override
+    public DataType[] inputTypes() {
+      return new DataType[] {DataTypes.IntegerType, DataTypes.StringType};
+    }
+
+    @Override
+    public DataType resultType() {
+      return DataTypes.StringType;
+    }
+
+    @Override
+    public String canonicalName() {
+      return "iceberg.truncate[width](string)";
+    }
+
+    @Override
+    public UTF8String produceResult(InternalRow input) {
+      Integer width = readAndValidateWidth(input);
+
+      UTF8String toTruncate = !input.isNullAt(1) ? input.getUTF8String(1) : null;
+      return toTruncate != null ? invoke(width, toTruncate) : null;
+    }
+  }
+
+  public static class TruncateBinary extends TruncateBase<byte[]> {
+    // magic method used in codegen
+    public static byte[] invoke(int width, byte[] value) {
+      if (value == null) {
+        return null;
+      }
+
+      return ByteBuffers.toByteArray(
+          TruncateUtil.truncateByteBuffer(width, ByteBuffer.wrap(value)));
+    }
+
+    @Override
+    public DataType[] inputTypes() {
+      return new DataType[] {DataTypes.IntegerType, DataTypes.BinaryType};
+    }
+
+    @Override
+    public DataType resultType() {
+      return DataTypes.BinaryType;
+    }
+
+    @Override
+    public String canonicalName() {
+      return "iceberg.truncate[width](binary)";
+    }
+
+    @Override
+    public byte[] produceResult(InternalRow input) {
+      Integer width = readAndValidateWidth(input);
+
+      byte[] toTruncate = !input.isNullAt(1) ? input.getBinary(1) : null;
+      return toTruncate != null ? invoke(width, toTruncate) : null;
+    }
+  }
+
+  public static class TruncateDecimal extends TruncateBase<Decimal> {
+    private final int precision;
+    private final int scale;
+
+    public TruncateDecimal(int precision, int scale) {
+      this.precision = precision;
+      this.scale = scale;
+    }
+
+    // magic method used in codegen
+    public static Decimal invoke(int width, Decimal value) {
+      if (value == null) {
+        return null;
+      }
+
+      return Decimal.apply(
+          TruncateUtil.truncateDecimal(BigInteger.valueOf(width), value.toJavaBigDecimal()));
+    }
+
+    @Override
+    public DataType[] inputTypes() {
+      return new DataType[] {DataTypes.IntegerType, DataTypes.createDecimalType(precision, scale)};
+    }
+
+    @Override
+    public DataType resultType() {
+      return DataTypes.createDecimalType(precision, scale);
+    }
+
+    @Override
+    public String canonicalName() {
+      return String.format("iceberg.truncate[width](decimal(%d,%d))", precision, scale);
+    }
+
+    @Override
+    public Decimal produceResult(InternalRow input) {
+      Integer width = readAndValidateWidth(input);
+
+      Decimal toTruncate = !input.isNullAt(1) ? input.getDecimal(1, precision, scale) : null;
+      return toTruncate != null ? invoke(width, toTruncate) : null;
+    }
+  }
+
+  private static Integer readAndValidateWidth(InternalRow input) {
```

Comment:
The logic here is odd. Why call `isNullAt` to box the value only to check whether it is null and throw an exception? I'd expect something like this:

```java
  if (input.isNullAt(0)) {
    throw IllegalArgumentException("Invalid truncation width: null");
  }

  return width;
```

I left out checking whether width is negative. I don't think we care.

Reply:
Removed the boxing.

For reference, the magic functions that take in only primitives will return `null` if the input width (an int) is `null`. But I think this is fine as `produceResult` isn't really called anyway.